pkg/repo/repo.go

```diff
@@ -0,0 +1,141 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package repo
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/theupdateframework/go-tuf"
+	"knative.dev/pkg/logging"
+)
+
+// CreateRepo creates and initializes a Tuf repo for Sigstore by adding
+// Fulcio Root Certificate, Rekor, and CTLog public keys into it.
+func CreateRepo(ctx context.Context, fulcio, rekor, ctlog []byte) (tuf.LocalStore, string, error) {
+	// TODO: Make this an in-memory fileystem.
+	tmpDir := os.TempDir()
+	dir := tmpDir + "tuf"
+	err := os.Mkdir(dir, os.ModePerm)
+	if err != nil {
+		logging.FromContext(ctx).Errorf("Failed to create tuf dir %s", err)
+		return nil, "", err
+	}
+	dir = dir + "/"
+	logging.FromContext(ctx).Infof("Creating the FS in %q", dir)
+	local := tuf.FileSystemStore(dir, nil)
+
+	// Create and commit a new TUF repo with the targets to the store.
+	logging.FromContext(ctx).Infof("Creating new repo in %q", dir)
+	//r, err := tuf.NewRepo(local)
+	r, err := tuf.NewRepoIndent(local, "", " ")
+	if err != nil {
+		logging.FromContext(ctx).Errorf("Failed to create NewRepo %s", err)
+		return nil, "", err
+	}
+
+	// Added by vaikas
+	if err := r.Init(false); err != nil {
+		logging.FromContext(ctx).Errorf("Failed to init repo %s", err)
+		return nil, "", err
+	}
+
+	// Make all metadata files expire in 6 months.
+	expires := time.Now().AddDate(0, 6, 0)
+
+	for _, role := range []string{"root", "targets", "snapshot", "timestamp"} {
+		_, err := r.GenKeyWithExpires(role, expires)
+		if err != nil {
+			logging.FromContext(ctx).Errorf("Failed to GenKeyWithExpires %s", err)
+			return nil, "", err
+		}
+	}
+
+	// This is the map of targets to add to the trust root with their custom metadata.
+	//	var targets map[string]json.RawMessage
+	if err := writeStagedTarget(dir, "rekor.pub", []byte(rekor)); err != nil {
+		logging.FromContext(ctx).Errorf("Failed to writeStagedTarget for rekor %s", err)
+		return nil, "", err
+	}
+	if err := writeStagedTarget(dir, "fulcio_v1.crt.pem", []byte(fulcio)); err != nil {
+		logging.FromContext(ctx).Errorf("Failed to writeStagedTarget for fulcio %s", err)
+		return nil, "", err
+	}
+	if err := writeStagedTarget(dir, "ctlog.pub", []byte(ctlog)); err != nil {
+		logging.FromContext(ctx).Errorf("Failed to writeStagedTarget for ctlog %s", err)
+		return nil, "", err
+	}
+
+	// Now add targets to the TUF repository.
+	// TODO(asraa): Targets does not get populated, so these never get added to
+	// repo. What should it contain?
+	// Looking at the gist that does work, it does AddTargets but without
+	// the expiry, so I changed to that since I don't know what I should be
+	// putting into the CustomMetadata below.
```

Comment:
yep! We can ignore custom metadata, so long as the filenames exactly match public root. See below comment.